sql: set planner database scope of SchemaChanger to descriptor db

otan opened the pull request:
Resolves #53819. Unsure about whether we should backport this.

Release note (bug fix): Previously, using builtins which referenced
other tables (e.g. nextval) in a CREATE TABLE AS or CREATE MATERIALIZED
VIEW would error. This has been resolved.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/54410)
<!-- Reviewable:end -->

otan commented:
hmm, this fails some tenant tests, but unsure why / how this happens

ajwerner commented:
This feels closely related to https://github.com/cockroachdb/cockroach/issues/51090

ajwerner commented:
which we did not successfully address in this 20.2 :(

thoszhang reviewed (commented):
Andrew said what I was going to say. What are the semantics here if the current database, the parent database of the table, and the parent database of the sequence aren't the same?

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/54410#-:-MHI2TIECSo3xxW4_Gy0:brxrusq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/schema_changer.go, line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/54410#-MHI0ZpNCwXOHgajEb61:-MHHuBzL79neKWh6nQDg:btuw6ee) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c68740b3d518c4220b894f933ca2118949a562d/pkg/sql/schema_changer.go#L256)):*
> ```Go
> 		// which relies on the planner scope being the same as the scope that ran the initial schema
> 		// query.
> 		db, err := catalogkv.GetDatabaseDescByID(ctx, txn, keys.SystemSQLCodec, table.GetParentID())
> ```

I think you need the codec on `sc.execCfg`. Also, you want to use `catalogkv.MustGetDatabaseDescByID()` here. `GetDatabaseDescByID` is sort of deprecated.


<!-- Sent from Reviewable.io -->

ajwerner commented:
I think the right thing to do is what’s proposed in the linked issue. We’ve done the hard part for 20.2 (make table ids stable). Now we can fix this, at least moving forward. It’s a little bit less clear what to do with existing sequences.

otan committed:
sql: set planner database scope of SchemaChanger to descriptor db

Release note (bug fix): Previously, using builtins which referenced
other tables (e.g. nextval) in a CREATE TABLE AS or CREATE MATERIALIZED
VIEW would error. This has been resolved.
